### DISCUSSION TITLE
macOS Big Sur 11.2.3 ld errors when building formula from source under Rosetta2

### DISCUSSION THREAD

**Discussion by Gcenx** [Homebrew/discussions#1021] [Everyday usage]
I'm working on a custom Formula and getting the following error during `make install`

```ld: dynamic main executables must link with libSystem.dylib for architecture x86_64```

However if I build this from Terminal manually I don't get any issues nor do I have any issues when building using Macports?

- Apple Silicon system, Terminal running under Rosetta2
- Homebrew was uninstalled and reinstalled
- Command Line Tools was manually removed and reinstalled using the 12.4 pkg
- Xcode 12.4 is also installed

**Comment by SMillerDev** [comment]
Are you depending on something that provides it's own `ld`? Try a run with hardcoded tool locations instead of depending on the path (since brew will add `depends_on` statements to the path)

**Reply by Gcenx** [reply]
It does require `mingw-w64` that could explain it, I’ll comment again once in able to check.

**Reply by carlocab** [reply]
That's likely to be a problem; `mingw-w64` does not work on ARM (yet).

**Reply by Gcenx** [reply]
> That's likely to be a problem; `mingw-w64` does not work on ARM (yet).

Please read my initial comment, I’m running this under Rosetta2 so that wouldn’t be the issue.

**Reply by carlocab** [reply]
Please read your error message.

>```
>ld: dynamic main executables must link with libSystem.dylib for architecture x86_64
>```

It sounds like `ld` doesn't think it's running under Rosetta2.

**Reply by SMillerDev** [reply]
`x86_64` is the architecture in Rosetta no?

**Reply by carlocab** [reply]
Yep. Sorry -- I didn't mean `ld`, but something in the build sounds like it's confused about being run in Rosetta.

**Reply by Gcenx** [reply]
> Please read your error message.
> 
> > ```
> > ld: dynamic main executables must link with libSystem.dylib for architecture x86_64
> > ```
> 
> It sounds like `ld` doesn't think it's running under Rosetta2.

The Terminal session brew install was ran from was launched under Rosetta2 so that’s not the issue.

From the same Terminal window I can build from source no issue.

> `x86_64` is the architecture in Rosetta no?

Yes it should be unless brew detect arch in some other manner.

**Reply by carlocab** [reply]
You wouldn't happen to have `coreutils` installed with the `gnubin` on your `PATH`, do you?

**Reply by Gcenx** [reply]
> You wouldn't happen to have `coreutils` installed with the `gnubin` on your `PATH`, do you?

If that were the cause (I don’t believe so) wouldn’t that have also broken builds outside of brew?

I didn’t explicitly install it and I don’t believe it’s a dependency of anything I’m using.

**Reply by carlocab** [reply]
It might break them; it might not. But if you don't remember doing so you probably didn't. Hard to say anything more about what's happening without knowing more about the formula at this point.

**Reply by Gcenx** [reply]
The problem was caused by homebrew shims, avoiding those gets the build to complete without issue.

**Reply by Bo98** [reply]
Given we have 5000+ formulae in homebrew-core using the shims without issue or exception, if you have more information on the issue I can perhaps look and see if there's a bug.

**Reply by Gcenx** [reply]
@Bo98 there are other formula that avoid homebrew shims, in this case it’s due to my running this under Rosetta2.

These shim scripts don’t account for being ran under Rosetta2, so when exec is called you’ll end up calling the arm64 slice instead of the intended x86_64 slice.

**Reply by Bo98** [reply]
> there are other formula that avoid homebrew shims

There shouldn't be any in homebrew-core (I'm not counting ones that do things to avoid them be used at _runtime_ post installation - but still use them at build-time). Let me know which and I'll fix them.

> These shim scripts don’t account for being ran under Rosetta2,

This is the first report I've seen from this. Rosetta 2 processes are supposed to launch subprocesses in Rosetta 2 automatically - I'll try test it out and see what's happening.

**Reply by Gcenx** [reply]
@Bo98 I didn’t read the formulas that avoided any shims too much just that others also did this as per @SMillerDev idea I’d set what I needed manually.

The shims I’d avoided were CC directly pointed to /usr/bin/clang and LD directly pointed to /usr/bin/ld, as these were set I also needed to set SDKROOT, with those set the formula builds, installs and runs without issue.

I’m more wondering how many users are trying to compile something from source when running under Rosetta2.

edit;
Make more sense
